### PR TITLE
Refactoring kimpy

### LINKED ISSUE #8: Pybind11 wrapper marked with py::nodelete and not deallocated on the C++ side
As per pybind11's documentation:

> Pybind11 provides a helper class py::nodelete that disables any destructor invocations. In this case, it is crucial that instances are deallocated on the C++ side to avoid memory leaks.

As far as I can tell, this is not what happens here and deallocation is pushed down to Python's realm where guarantees are really limited (`__init__` and `__del__` can be called any number of time, **including not at all**). This could very easily lead to memory leaks and use-after-free errors.

The correct way to handle this is either by defining a wrapping C++ class or changing the custom deleter in `unique_ptr`. I suggest simply using the latter.

------

If we take `https://github.com/openkim/kimpy/blob/master/kimpy/KIM_Collections_bind.cpp` as an example, I believe the following would do what we are expecting (but please thoroughly test it, don't take my word for it):

```
py::class_<Collections, std::unique_ptr<Collections, decltype(&Collections::Destroy)> > cl(
      module, "Collections");
...
```

Note how `py:nodelete` got transformed into `decltype(&Collections::Destroy)`.

### PR DESCRIPTION
**Pull request by yafshar**
### Summary

This PR removes the manual memory management (where `kimpy` users had to explicitly create and destroy objects)

- Removes the manual memory management. This update addresses the issue https://github.com/openkim/kimpy/issues/8
  
- Update the implementation. Now there is no `error` message returned back from the `KIM-API` to the users. In case of an error, it throws a `RuntimeError` exception with a message explaining the reason.
  For example, you can create a `collections` object as,

  ```py
  collections = kimpy.collections.create()  # collections = kimpy.collections.Collections()
  ```

  or,

  ```py
  try:
    collections = kimpy.collections.create()
  except RuntimeError:
    msg = 'Calling "kimpy.collections.create" failed.'
    raise kimpy.KimPyError(msg)
  ```

- Update the returned values and input arguments naming to lower case.
- Add document to all methods
- Update the NeighList bindings.
   Now to use the neighlist, 
   ```py
   # create a NeighList object 
   neigh = neighlist.create()    # or neigh = neighlist.NeighList()

   # build the neighbor list
   neigh.build(coords, influence_dist, cutoffs, need_neigh)
   
   # get number of neighbors and neighbors of the requested particle
   number_of_neighbors, neighbors_of_particle = neigh.get_neigh(cutoffs, neigh_list_index, particle)
   ```

### Related Issue(s)

https://github.com/openkim/kimpy/issues/8

### Backward Compatibility

This PR is not backward compatible

### Author(s)

Yaser Afshar (UMN)

### CODE REVIEW

**Comment by yafshar** [comment]
@mjwen is there any reason not to put the methods `nbl_create_paddings, nbl_build, nbl_get_neigh, nbl_clean` in the `NeighList` class? Also, update some of the namings?

**Comment by ellio167** [comment]
@yafshar There's lots of changes here, so I won't be able to do a comprehensive/detailed review.  I'm broadly happy with this.  I think supporting python's exception handling is a good thing.  I'm not sure it is worth adding the pickling stuff at this point.  I would probably argue to remove it from this work and include this later if/when a complete pickling solution is worked out.

@mjwen I would want to have your approval for these various changes.  I think they are moving in the right direction, but you know all this much more intimately than I do.

**Comment by mjwen** [comment]
Thanks @yafshar, I agree with @ellio167 that this is very enhancement!

Given that some parts of the kim-api is not easily picklable (see #7), let's remove the pickling part? Removal of the manual memory management is definite what we need to do, but I have concerns on the error code part: backward compatibility is broken. 

Can we do this more gradually by first giving a version with deprecation warnings? Based on the Used-By statistics, there are 16 other projects using this, and I am afraid it's not a good idea to make this abrupt change. What do you think of this @ellio167 ?

I'll go through the files once we've decided what we want.

**Comment by mjwen** [comment]
> @mjwen is there any reason not to put the methods `nbl_create_paddings, nbl_build, nbl_get_neigh, nbl_clean` in the `NeighList` class? Also, update some of the namings?

The neighbor list code was ported from some old openkim-python utility, and I kept the layout. I don't think there is any stuff preventing the NeighList struct and the other functions being combined into one class.

**Comment by ellio167** [comment]
> Thanks @yafshar, I agree with @ellio167 that this is very enhancement!
> 
> Given that some parts of the kim-api is not easily picklable (see #7), let's remove the pickling part? Removal of the manual memory management is definite what we need to do, but I have concerns on the error code part: backward compatibility is broken.
> 
> Can we do this more gradually by first giving a version with deprecation warnings? Based on the Used-By statistics, there are 16 other projects using this, and I am afraid it's not a good idea to make this abrupt change. What do you think of this @ellio167 ?
> 
> I'll go through the files once we've decided what we want.

If we increment the major version number, then the backward incompatibility is ok, in principal.  We could do a release with deprecation warnings, but unless we are going to wait some time before releasing this version, then I’m not sure it is all that advantageous. Probably we can just go directly to the new major version release and include clear warnings in the release notes about the backward incompatibility...

File: tests/test_sem_ver.py [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
 def test_main():
-
     version = kimpy.sem_ver.get_sem_ver()
-    ls_less_than, error = kimpy.sem_ver.is_less_than('2.0.1', '2.1.0')
-    check_error(error, 'kimpy.sem_ver.is_less_than')
-    out = kimpy.sem_ver.parse_sem_ver(version)
-    version_related = out[:-1]
+
+    try:
+        ls_less_than = kimpy.sem_ver.is_less_than('2.0.1', '2.1.0')
```
Comment (mjwen) [inline tests/test_sem_ver.py]:
```suggestion
        is_less_than = kimpy.sem_ver.is_less_than('2.0.1', '2.1.0')
```

File: tests/test_sem_ver.py [hunk 2]
```diff
@@ -1,18 +1,26 @@
...
+    except RuntimeError:
+        msg = 'Calling "kimpy.sem_ver.is_less_than" failed.'
+        raise kimpy.KimPyError(msg)
+
+    assert ls_less_than == 1
```
Comment (mjwen) [inline tests/test_sem_ver.py]:
```suggestion
    assert is_less_than == 1
```

File: kimpy/neighlist/neighbor_list_bind.cpp [hunk 3]
```diff
@@ -8,116 +8,146 @@
 #include <memory>
 #include <vector>
```
Comment (mjwen) [inline kimpy/neighlist/neighbor_list_bind.cpp:7]: `neighbor_list.h` is not included. Can you reminder me how this works?  (Haven't worked on C++ for a while)

File: kimpy/py_kim_wrapper.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * to the user and it is hidden behind a forwarding facade. This class has a
+ * private constructor & destructor. It is also creating a new
+ * \c ComputeArguments object for the Model object.
+ *
+ * The naive bindings via pybind11 and using ``py::nodelete`` class requires
+ * users to manullay create and destroy the objects. Which could very easily
+ * lead to memory leaks/use-after-free (segmentation fault) errors.
+ *
+ * By using the current wrapper class we remove the need for user requirement
+ * to explicitely edetrsoy the object. Also provide a information that allows
```
Comment (mjwen) [inline kimpy/py_kim_wrapper.h]:
```suggestion
 * to explicitly destroy the object. Also provide information that allows
```

File: tests/test_callbacks.py [hunk 5]
```diff
@@ -1,19 +1,21 @@
...
+we test the data modified in the KIM KPI is reflected locally by changing the
+data in the callbacks when it is last called.
+
 """
 
 import numpy as np
 import kimpy
-from kimpy import check_error, report_error
 from ase import Atoms
 import pytest
```
Comment (mjwen) [inline tests/test_callbacks.py:21]: Remove `assert_1d_array()`  and `assert_2d_array`; directly use `np.allclose(rtol=..., atol=...)`?

File: scripts/KIM_FieldName_bind.cpp-template [hunk 6]
```diff
@@ -8,52 +8,74 @@
...
   cl.def(py::init<>())
     .def(py::init<int const>())
     .def(py::init<std::string const>())
     .def("known", &FieldName::Known)
     .def(py::self == py::self)
     .def(py::self != py::self)
-    .def("__repr__", &FieldName::ToString);
-
+    .def("__repr__", &FieldName::ToString)
+    .def(py::pickle(
```
Comment (mjwen) [inline scripts/KIM_FieldName_bind.cpp-template]: Remove this?

**Comment by mjwen** [comment]
@yafshar I went over most of the files. But for some of the heavily modified ones, it's difficult to see the modifications. Can you push a new commit to format the code by running this https://github.com/openkim/kimpy/blob/master/utils/format_sources.py

**Comment by yafshar** [comment]
> @yafshar I went over most of the files. But for some of the heavily modified ones, it's difficult to see the modifications. Can you push a new commit to format the code by running this https://github.com/openkim/kimpy/blob/master/utils/format_sources.py

@mjwen I generally tried to follow the `clang-format` format. Since using `format_sources.py` script will cause problems especially at places when I used `R"pbdoc` for documentation. Do you know a way to customize the formatter.

**Comment by yafshar** [comment]
@mjwen also this script is intrusive and I had no idea :). I ran it in the main root of `kimpy` and since it is using `..` path it went to my other repositories and now I ended up with a mess for lots of reformatted files based on `clang-format`.

**Comment by yafshar** [comment]
@mjwen just an example issue using this formatted for python code,

```py
def dirs_for_collection(collection, collections):
    for it in (kimpy.collection_item_type.modelDriver,
               kimpy.collection_item_type.portableModel,
               kimpy.collection_item_type.simulatorModel):
        try:
            extent = collections.cache_list_of_directory_names(collection, it)
        except RuntimeError:
            msg = 'Calling "collections.cache_list_of_directory_names" failed.'
            raise kimpy.KimPyError(msg)

        print('{}:{}:'.format(str(collection), str(it)))

        for i in range(extent):
            try:
                name = collections.get_directory_name(i)
            except RuntimeError:
                msg = 'Calling "collections.get_directory_name" for index = '
                msg += '{} failed.'.format(i)
                raise kimpy.KimPyError(msg)

            print('    ', name)
```

after using the formmatter,

```py
    def dirs_for_collection(collection, collections) : for it in(kimpy.collection_item_type.modelDriver, kimpy.collection_item_type.portableModel, kimpy.collection_item_type.simulatorModel):
try : extent = collections.cache_list_of_directory_names(collection, it)
        except RuntimeError:
            msg = 'Calling "collections.cache_list_of_directory_names" failed.'
            raise kimpy.KimPyError(msg)

        print('{}:{}:'.format(str(collection), str(it)))

        for i in range(extent):
  try : name = collections.get_directory_name(i)
            except RuntimeError:
                msg = 'Calling "collections.get_directory_name" for index = '
                msg += '{} failed.'.format(i)
                raise kimpy.KimPyError(msg)

            print('    ', name)
```

using this formatted for cpp code `KIM,

```cpp

```

```cpp
  module.def(
      "create",
      []() {
        Collections * collections;

        int error = Collections::Create(&collections);
        if (error == 1)
        {
          throw std::runtime_error(
              "Unable to create a new KIM-API collections object!");
        }

        return std::unique_ptr<Collections, PyCollectionsDestroy>(
            std::move(collections));
      },
      R "pbdoc(
          Create a new KIM
          - API collections object.

            Returns:Collections
      : collections) pbdoc ");
```

**Comment by yafshar** [comment]
> @yafshar I went over most of the files. But for some of the heavily modified ones, it's difficult to see the modifications. Can you push a new commit to format the code by running this https://github.com/openkim/kimpy/blob/master/utils/format_sources.py

@mjwen OK, I figured it out. and reformatted all the files using the script. Also did a minor fix in the script to only change the files within kimpy repository.

**Comment by yafshar** [comment]
> @yafshar There's lots of changes here, so I won't be able to do a comprehensive/detailed review. I'm broadly happy with this. I think supporting python's exception handling is a good thing. I'm not sure it is worth adding the pickling stuff at this point. I would probably argue to remove it from this work and include this later if/when a complete pickling solution is worked out.

@ellio167 @mjwen I removed the pickling part from this PR.

**Comment by yafshar** [comment]
@mjwen I did a minor update on the `NeighList` binding so it can be used in a more natural way, for example,

```py
# create a neighbor list
neigh = neighlist.create()

# or
neigh = neighlist.NeighList()

# you can use both ways
neighlist.build(neigh, coords, influence_dist, cutoffs, need_neigh)

# or 
neigh.build(coords, influence_dist, cutoffs, need_neigh)


number_of_neighbors, neighbors_of_particle = neighlist.get_neigh(neigh, cutoffs, neigh_list_index, particle)

# or

number_of_neighbors, neighbors_of_particle = neigh.get_neigh(cutoffs, neigh_list_index, particle)
```

I think this is natural to the users, if you agree and we go with this PR and do this major, not backward compatible update, it would be better to discard the old way and keep the second approach. Please let me know what do you think?

**Review by mjwen — COMMENTED** [review]
> @mjwen also this script is intrusive and I had no idea :). I ran it in the main root of `kimpy` and since it is using `..` path it went to my other repositories and now I ended up with a mess for lots of reformatted files based on `clang-format`.

Sorry to hear that. It is supposed to be run in the directory where it resides...

**Comment by mjwen** [comment]
> @mjwen I did a minor update on the `NeighList` binding so it can be used in a more natural way, for example,
> 
> ```python
> # create a neighbor list
> neigh = neighlist.create()
> 
> # or
> neigh = neighlist.NeighList()
> 
> # you can use both ways
> neighlist.build(neigh, coords, influence_dist, cutoffs, need_neigh)
> 
> # or 
> neigh.build(coords, influence_dist, cutoffs, need_neigh)
> 
> 
> number_of_neighbors, neighbors_of_particle = neighlist.get_neigh(neigh, cutoffs, neigh_list_index, particle)
> 
> # or
> 
> number_of_neighbors, neighbors_of_particle = neigh.get_neigh(cutoffs, neigh_list_index, particle)
> ```
> 
> I think this is natural to the users, if you agree and we go with this PR and do this major, not backward compatible update, it would be better to discard the old way and keep the second approach. Please let me know what do you think?

Totally agree that the latter class based one is more natural. Please do it!

**Comment by mjwen** [comment]
> > @yafshar There's lots of changes here, so I won't be able to do a comprehensive/detailed review. I'm broadly happy with this. I think supporting python's exception handling is a good thing. I'm not sure it is worth adding the pickling stuff at this point. I would probably argue to remove it from this work and include this later if/when a complete pickling solution is worked out.
> 
> @ellio167 @mjwen I removed the pickling part from this PR.

Thanks!

File: kimpy/neighlist/neighbor_list_bind.cpp (hunk 3 above)
Comment (yafshar) [reply kimpy/neighlist/neighbor_list_bind.cpp:7]: It has the header, and `neighbor_list.h` is included, but I assume formmater changed the order and put it at the top. 

https://github.com/openkim/kimpy/blob/b0ce7162291044361ee7da8d0d3c7d598346e33d/kimpy/neighlist/neighbor_list.cpp#L3

**Comment by yafshar** [comment]
@mjwen If we are going to this major backward incompatible update, Ellad suggested emailing/informing all the projects using kimpy about this update and ask if they have limitations or constraints to update. I can also provide help on updating. Do you know those projects? (can you provide the name of those? or should I check the GitHub info on projects which are using kimpy?)

**Comment by mjwen** [comment]
> should I check the GitHub info on projects which are using kimpy?)

Please do it. I do not have other info than this.

**Comment by yafshar** [comment]
> > should I check the GitHub info on projects which are using kimpy?)
> 
> Please do it. I do not have other info than this.

I checked all these projects and they are all using ASE and none of them is using kimpy directly. I do not think this PR would cause any issue in these projects. Me and Dan have already updated ASE with this PR and it seems working perfectly fine.

**Comment by mjwen** [comment]
@yafshar All seem good to me and I am merging it into master. Can you please prepare a ChangeLog item for this (edit CHANGELOG.txt directly)? 

Also, I am going to update my fitting code that uses kimpy and do some tests there to make sure everything is behaving as expected. 

After all this, we can make a major release. Thanks again for the PR -- a huge enhancement!

**Pull request merged**